### ISSUE TITLE
Use `opt_einsum` for the biaffine layer

### ISSUE THREAD

**Issue by LoicGrobol** [hopsparser/npdependency#60]
See <https://optimized-einsum.readthedocs.io>, in the most basic way this could be a drop-in, but there are things that we know are constant (input and output dim), when the parser is in eval mode, the weight tensor is constant and when training or parsing on a dataset the batch size is constant in our current setting (except maybe for the last batch).

It might be worth it to wait until #33, or at least a speed benchmark, to see if the gain in perf is worth the added dependency.

**Comment by LoicGrobol** [comment]
[Actually](https://pytorch.org/docs/stable/generated/torch.einsum.html) even just installing `opt_einsum` already allows torch to do some optimizations transparently so start with that!
